### ISSUE TITLE
Empty Support phone number check does not run without reloading the page

### ISSUE THREAD

**Issue by leonardola** [Automattic/woocommerce-payments#6672]
### Describe the bug
After you set a support phone number you're not supposed to be able to remove it. There is a check [here](https://github.com/Automattic/woocommerce-payments/blob/e14ff86eaf0f5da634d4dc86632ce77501793ea5/client/settings/support-phone-input/index.js#L31-L36) but if the merchant sets the phone up, save and remove the phone the check does not run correctly. The phone is not saved but no warning is shown to the merchant. I haven't tested but this might be happening for other fields too.

This is not a duplicate of #5516 This bug is that we only start to check if the field is empty after the page is reloaded not right after the settings are saved with a new phone.

### To Reproduce
- Go to the merchant site \wp-admin > Payments > Settings > Transaction preferences
- Make sure you don't have a support phone number set
- Add a valid phone number
- Click save
- Remove the phone number 
- Click save
- The check did not run. The empty phone was not saved but no error showed up

### Actual behavior
The check should run correctly for newly added data.

**Comment by htdat** [comment]
It seems a duplicate of #5516

@leonardola - I am going to close it, feel free to re-open this issue in case it's not the same one.

**Comment by leonardola** [comment]
It's not exactly the same error. #5516 is about not being able to save a number. This one is failing to validate that the number field is empty before saving. 

When the merchant visits the settings page for the first time there is no phone number so when they save the changes we don't check for an empty phone number field as one was never inserted and it's ok to not have one in place. As soon as the merchant inserts a phone number stripe does not allow a merchant to remove it only update it. So we should start to check if the field is empty. The bug is that we only start to check if the field is empty after the page is reloaded not right after the settings are saved.

**Comment by htdat** [comment]
> The bug is that we only start to check if the field is empty after the page is reloaded not right after the settings are saved.

@leonardola - Aha, that makes sense. Thanks for the details. Could you update this part into `Describe the bug` or `To Reproduce` section above? 

> There is a check [here](https://github.com/Automattic/woocommerce-payments/blob/add%2Fwarning-to-advanced-settings/client/settings/support-phone-input/index.js#L31-L36)

Would be nice to update this link to a permalink one too. It's now showing 404.

**Comment by htdat** [comment]
@leonardola has already updated this issue. Moving this back to the prioritization progress for team Gamma.

**Comment by jessy-p** [comment]
I tested this with the latest code (before merging #7851) and it appears to be working now. @htdat could you please confirm.

**Comment by htdat** [comment]
@jessy-p  No, the issue is still there. The first tricky part is this step to reproduce: 

> Make sure you don’t have a support phone number set. 

You might either create a new merchant account or hack around to have this state. The next tricky part is that you would not refresh the settings page. 
In other words, the issue will happen for only the first time. But I think it's pretty rare that right after adding a valid phone, merchants set the phone number back to empty.

The main issue is this line where `currentPhone` is persisted during the lifetime of the component: 
https://github.com/Automattic/woocommerce-payments/blob/f59a0b27c034bb4a3cfda83021a62dd8c0754ff4/client/settings/support-phone-input/index.js#L20

**Comment by jessy-p** [comment]
This issue impacts Payments settings page, so assigning to Moltres (based on team responsibilities Pc2DNy-3z-p2) @daquinons. Assigning as part of Gamma Triage process PcreKM-yM-p2.

**Comment by elazzabi** [comment]
Howdy 👋 

I looked at this and tried a few things. Unfortunately, there is no direct and clean way of doing it.

To summarize the need:
**We want the PhoneInput component to "reload" after saving the settings.**

The solution:
**Given the save happens outside the component, our only way is to rely on the Redux store.**

The problem:
With the `useSettings()` hook, we can access `isSaving` and `invalidateResolutionForStoreSelector( 'getSettings' )` if the settings are not saving. The problem is `isSaving` starts as false, and this will create an infinite loop.

Then I tried to add another check, have we changed the phone input? The problem here is that it gets executed after every keystroke, not fun.

Finally, I think the best way is to keep track of the saving status and update the component if that changes. Something like:

```js
const { isSaving } = useSettings();
const { invalidateResolutionForStoreSelector } = useDispatch( STORE_NAME );
const wasSaving = useRef( isSaving );

...

// Refresh settings when saving completes
useEffect( () => {
	if ( wasSaving.current && ! isSaving ) {
		invalidateResolutionForStoreSelector( 'getSettings' );
	}
	wasSaving.current = isSaving;
}, [ isSaving, invalidateResolutionForStoreSelector ] );
```

Summary:
- Given the bug described in this issue, I don't believe it's worth the effort. The solution seems "hacky" and forced.
- I believe this is a problem with all inputs, but PhoneInput is the only one where we have this requirement (can't be empty if set before)

So back to you @htdat @jessy-p @leonardola because you are the ones active in this issue 😅 Do we want to go with this hacky way? Or do you have any other opinions?

Thanks

**Comment by htdat** [comment]
@elazzabi 

> Then I tried to add another check, have we changed the phone input? The problem here is that it gets executed after every keystroke, not fun.

If this is limited to `SupportPhoneInput`, I think that's fine. Actually, the whole validation of `SupportPhoneInput` also relies on each change of the input to provide the validation.

**Comment by elazzabi** [comment]
Yes, this is limited to `SupportPhoneInput` as far as I can see. And I agree it's safe to "ignore" this issue for now.
